userapi/src/main/java/model/D42Model.java

```diff
@@ -0,0 +1,117 @@
+    }
+
+    private Set<String> getListingAtLevelWithPrefix(String level, String objPrefix, boolean encoded) throws Exception {
+        Set<String> commonPrefixes = store.listAsync(level, false).get(TIMEOUT, TimeUnit.SECONDS);
+        Set<String> objs = new HashSet<>();
+        for (String commonPrefix : commonPrefixes) {
+            String objName = getLastFromCommonPrefix(commonPrefix);
+            if (encoded) {
+                objName = new String(base32.decode(objName.getBytes()));
+            }
+            if (objName.indexOf(objPrefix) == 0) {
+                objs.add(objName);
+            }
+        }
+        return objs;
+    }
+
+    @Override
+    public void setStorage(AsyncStorage store) {
+        this.store = (S3AsyncStorage) store;
+    }
+
+    @Override
+    public Set<String> getAppIdsWithPrefix(String appIdPrefix) throws Exception {
+        return getListingAtLevelWithPrefix(BUCKET + PATH_SEPARATOR + VERSION + DELIMITER, appIdPrefix, true);
+    }
+
+    @Override
+    public Set<String> getClusterIdsWithPrefix(String appId, String clusterIdPrefix) throws Exception {
+        return getListingAtLevelWithPrefix(
+                BUCKET + PATH_SEPARATOR + VERSION + DELIMITER + new String(base32.encode(appId.getBytes())) + DELIMITER,
+                clusterIdPrefix,
+                true);
+    }
+
+    @Override
+    public Set<String> getProcsWithPrefix(String appId, String clusterId, String procPrefix) throws Exception {
+        return getListingAtLevelWithPrefix(BUCKET + PATH_SEPARATOR + VERSION + DELIMITER + new String(base32.encode(appId.getBytes()))
+                        + DELIMITER + new String(base32.encode(clusterId.getBytes())) + DELIMITER,
+                procPrefix,
+                false);
+    }
+
+    @Override
+    public Set<Profile> getProfilesInTimeWindow(String appId, String clusterId, String proc, String startTime, String durationInSeconds) throws Exception {
+        String prefix = BUCKET + PATH_SEPARATOR + VERSION + DELIMITER + new String(base32.encode(appId.getBytes()))
+                + DELIMITER + new String(base32.encode(clusterId.getBytes())) + DELIMITER + proc + DELIMITER;
+        Map<Profile, Set<String>> profilesMap = store.listAsync(prefix, true).get().stream().collect(Collectors.groupingBy(D42Model::getProfile, Collectors.mapping(D42Model::getWorkType, Collectors.toSet())));
+        Set<Profile> profiles = profilesMap.entrySet().stream().map(profileSetEntry -> {
+            profileSetEntry.getKey().setValues(profileSetEntry.getValue());
+            return profileSetEntry.getKey();
+        }).collect(Collectors.toSet());
+        try {
+            ZonedDateTime startTimeAsTime = ZonedDateTime.parse(startTime);
+            Long durationInSecondsAsLong = Long.parseLong(durationInSeconds);
+            profiles = profiles.stream().filter(k -> ZonedDateTime.parse(k.getEnd()).isBefore(startTimeAsTime.plusSeconds(durationInSecondsAsLong).plusNanos(1))
```

Comment:
so the time window here is for the startTime of the stored profiles.
Basically, if the window given is : 11 am with 30 min duration, then we should return all profiles which were started in this window.

Also, i think its better to add the comment in case of long stream operations.